### PR TITLE
template command update, adding in versions

### PR DESCRIPTION
**Pull request by rhyeal**
I realized the Homebrew file was autogenerated. It might be time to think about a better way to structure this file setup with templated files.

Also includes the version being replaced in the MacOS file on install now

### CODE REVIEW

**Comment by mmrwoods** [comment]
@rhyeal Would it not be better to remove the template? 

It seems it's only need to automatically update the formula url and sha from CircleCI, but that comes at the cost of maintaining two copies of the Homebrew formula file, or at least at the cost of confusion for anyone contributing needing to know they should edit the template file rather than the formula file.

How about removing the template and just using `sed -i` to update the url and sha in the formula file from CircleCI? This seems a lot simpler than maintaining a separate template file, setting env vars in CircleCI, injecting them into a template using `envsubst`, and then overwriting the formula with the output generated from the template.

Something similar could be done with the README, again it seems to only require a template to inject the version, shas, and checksums, but these are all within one section which could be swapped out using `sed` and `awk`. Or, maybe the checksums could just be moved to a different file entirely, and that could be generated from CircleCI (with the template used to generate it stored in the `.circleci` directory to avoid confusion?

File: aws-rotate-iam-keys.template.rb [hunk 1]
```diff
@@ -21,11 +21,15 @@ class AwsRotateIamKeys < Formula
...
-      version %x{git describe develop --always | tr -d '\n'}
+      version "head"
     end
   end
 
   def install
+    version = File.read('VERSION')
+    code = File.read("src/bin/aws-rotate-iam-keys")
+    code = code.gsub(/<%VERSION%>/, version)
+    File.open("src/bin/aws-rotate-iam-keys", "w") {|file| file.puts code}
```
Comment (mmrwoods) [inline aws-rotate-iam-keys.template.rb:32]: I haven't tested this, but I don't think it'll work as the current working directory when this code is run won't necessarily be the directory containing the `VERSION` file.

**Comment by mmrwoods** [comment]
@rhyeal Sorry, I've just had a glance at this again, and I've realised that the intention with the templates is that we should only be editing the template files, never the actual homebrew formula and readme files.

I think it makes contributing to the project cumbersome though, as prior knowledge about this unusual setup is required (I know there is a note at the end of the readme, but, well, it's at the end ;-)

The templates seem to be used solely to include version numbers and checksums in the homebrew formula file, the readme, and the website html. There must be a way to improve this such that it does not otherwise interfere with contributions to the source. 

Changes to the website source are not that big a deal because there is no `index.html` in the repo, just the `index.template.html`, so it's pretty clear what's going on there, but the templates for the homebrew formula and readme are source that generates source, which is a bit confusing. 

Maybe it wouldn't be confusing if the templates were moved somewhere other that the root, and the generated files had a big notice at the top rather than at the bottom? 

I still think I'd be looking for find a way to achieve the same end result without using templates at all though. It would be nice if this generation of version numbers and checksums and injection of those into the source was isolated as part of a build process, and was not effectively part of the general source code.

File: aws-rotate-iam-keys.template.rb (hunk 1 above)
Comment (rhyeal) [reply aws-rotate-iam-keys.template.rb:32]: Gotcha. I'll re-work this section and test it my different systems.

**Comment by rhyeal** [comment]
> @rhyeal Would it not be better to remove the template?


I was thinking about this after the commit last night. I agree that removing the templates and moving towards a `sed` or `awk` replacement is a much better fix.

File: aws-rotate-iam-keys.template.rb (hunk 1 above)
Comment (mmrwoods) [reply aws-rotate-iam-keys.template.rb:32]: I should have also asked what the purpose of this is. It looks like more magic/voodoo to rewrite source code on the fly, in this case to inject the correct version number into the script when it is installed via homebrew, but this is all so hard to follow. It works, but at the cost of making things very hard to understand - where is the true source?

Comment (rhyeal) [reply aws-rotate-iam-keys.template.rb:32]: The source for this is the zipped files. There are numerous release artifacts in different formats that come out of this repo each time a new version is cut, including the state of the repo itself for the Homebrew installation. I'm considering scrapping the template stuff as well as this code change in favor of a folder change that puts all the source files in `src` and then the resulting files will be built to `Formula` for homebrew, `releases` for the deb and zip files, and `website` for the website. I think that will clear up any confusion on the code files. Then I'll come up with a `sed` replace for the README to get rid of that template. Thoughts?

Comment (mmrwoods) [reply aws-rotate-iam-keys.template.rb:32]: I don't think any build process should generate source files, it should just package them. 

Does the Homebrew archive really need to include the entire repo? I would have thought it just needs to contain the files needed to install the binary. 

The Homebrew formula is a source of confusion, because having a Homebrew formula included in the same repo as the binary it installs is unusual. However, I don't think you should try and change that just yet, small steps and all. If the source of the formula moves out of `Formula`, but there is still a formula file in `Formula`, that's going to be confusing.

I think it would less confusing to move the `aws-rotate-iam-keys` script into a `bin` directory from the root, create a new `debian` directory at the root for the debian stuff, delete the `src` directory and then treat all these files as actual source rather than build templates. Then you'd have a project root with `Formula`, `Windows`, `bin`, `debian` and `website`, which seems pretty clear to me.

I might be simplifying, but building a release without templates seems like it should be easy enough... bump the version, generate the release archives, update the various checksums.

I don't know anything about debian packages, so maybe it makes complete sense for the files in the `debian` directory to be templates with an interpolated `VERSION` variable for building the debian/ubuntu packages. If templates aren't necessary, it seems like it would be simple to replace the use of build templates with `sed` (the `VERSION` variable is only used to set the `Version` and `Standard-Version` fields in the debian and ubuntu control files - p.s. should this really be used to set `Standard-Version`?).

Bumping the version in Homebrew formula should be simple using `sed`, and could be simple in the `aws-rotate-iam-keys` script itself by setting a `VERSION` variable at the top of the script rather than interpolating the `VERSION` into the help output. Updating the readme is admittedly a little trickier, but surely not that hard (and could be really simple if it was updated slightly to just avoid referencing specific version numbers).

Generating the release archives seems to work fine already, though the Homebrew archive doesn't need the entire repo, just the bin directory.

Updating the checksums is currently tricky because this also updates the readme, but it doesn't have to, those checksums could be separate files in the `releases` directory, one for each archive/package, and bumping the sha in the homebrew formula is easy with `sed`.

Sorry, gotta go, sorry if some of the above is long winded, but haven't time to summarise in a TLDR

File: aws-rotate-iam-keys.template.rb [hunk 2]
```diff
@@ -21,11 +21,15 @@ class AwsRotateIamKeys < Formula
   devel do
     Dir.chdir(File.expand_path(File.join(File.dirname(__FILE__), '../'))) do
       url %x{git config --local --get remote.origin.url | tr -d '\n'}, using: :git, branch: "develop"
-      version %x{git describe develop --always | tr -d '\n'}
+      version "head"
     end
   end
```
Comment (bcox24) [inline aws-rotate-iam-keys.template.rb:26]: should this devel block just be removed because of issue #43?

Comment (mmrwoods) [reply aws-rotate-iam-keys.template.rb:26]: Yep, it definitely should be removed, if the code is ever merged. I'm not sure the code in this PR is destined for a merge anymore though, at least as it stands. The PR seems to have turned more into a discussion about how to simplify the build process and to make contributing easier.

How to deal with the deprecated devel blocks is a tricky one though, there are forks of this project using the devel software spec to allow simple installation of the latest development code from the fork for real world testing before pull requests are submitted back to the upstream project. That's both my fault, and my problem, I'll need to figure something out pronto, because the deprecation warning is a pain in the ass.

**Comment by mmrwoods** [comment]
Closing as no longer relevant, build process is failing and most template have been removed. If a build process is revived, it should be simplified to avoid using templates unnecessarily.
